src/common/models/visualization-props/visualization-props.ts

```diff
@@ -33,8 +32,32 @@ export interface VisualizationProps {
   isThumbnail?: boolean;
 }
 
-export interface DatasetLoad {
-  loading?: boolean;
-  dataset?: Dataset;
-  error?: any;
+enum DatasetLoadStatus { LOADED, LOADING, ERROR }
+
+interface DatasetLoadBase {
+  status: DatasetLoadStatus;
+}
+
+interface DatasetLoading extends DatasetLoadBase {
+  status: DatasetLoadStatus.LOADING;
+}
+
+interface DatasetLoaded extends DatasetLoadBase {
+  status: DatasetLoadStatus.LOADED;
+  dataset: Dataset;
 }
+
+interface DatasetLoadError extends DatasetLoadBase {
+  status: DatasetLoadStatus.ERROR;
+  error: Error;
+}
+
+export const loading: DatasetLoading = { status: DatasetLoadStatus.LOADING };
+export const error = (error: Error): DatasetLoadError => ({ error, status: DatasetLoadStatus.ERROR });
+export const loaded = (dataset: Dataset): DatasetLoaded => ({ status: DatasetLoadStatus.LOADED, dataset });
+
+export const isLoading = (dl: DatasetLoad): dl is DatasetLoading => dl.status === DatasetLoadStatus.LOADING;
```

Comment:
This is very functional but I would like regular polymorphism as long as there is a class hierarchy. DataLoadBase could define all isFoo methods as "return false" and then specific implementation could overwrite the behaviour. 
Finally the call looks like datasetLoading.isLoading() instead isLoading(datasetLoading).
In addition each implementation could define methods for valid state machine transitions, or throw an exception if the transition is forbidden.

Reply:
convinced, thanks!